stm32/libcanard/bxcan/src/bxcan.c

```diff
@@ -0,0 +1,753 @@
+                //  1. 0x123        <-- Takes mailbox 0 (or any other)
+                //  2. 0x122        <-- Takes mailbox 2 (or any other)
+                //  3. 0x121        <-- Takes mailbox 1 (or any other)
+                //  4. 0x120        <-- INNER PRIORITY INVERSION HERE (only if all three TX mailboxes are still busy)
+                // This situation is even less likely to cause any noticeable disruptions on the CAN bus. Despite that,
+                // it is better to warn the developer about that during debugging, so we fire an assertion failure here.
+                // It is perfectly safe to remove it.
+                // Note: this error condition (all TX mailboxes busy) also occurs when erroneously sending data to the
+                // TX mailboxes of a non-existing CAN interface. Eg: accessing CAN2 on a device that only has CAN1.
+#if BXCAN_CONFIG_DEBUG_INNER_PRIORITY_INVERSION
+                BXCAN_ASSERT(!"CAN PRIO INV");  // NOLINT
+#endif
+                prio_inversion = true;
+            }
+            else
+            {
+                prio_inversion = false;
+            }
+        }
+        else  // All TX mailboxes are free, use first
+        {
+            tx_mailbox = 0U;
+        }
+
+        BXCAN_ASSERT(tx_mailbox < 3U);  // Index check - the value must be correct here
+    }
+
+    // By this time we've proved that a priority inversion would not occur, and we've also
+    // found a free TX mailbox. Therefore it is safe to enqueue the frame now.
+    if ((input_ok == true) && (prio_inversion == false))
+    {
+        volatile BxCANTxMailboxType* const mb = &bxcan_base->TxMailbox[tx_mailbox];
+
+        // The payload can be any length up to the MTU. We first make a local copy of the
+        // available payload bytes in a zero-filled array of MTU size. This makes the logic
+        // for filling the mailbox simpler as the payload pointer can be to variable-size
+        // data or even be a NULL pointer.
+        static uint8_t scratch_data[8] = {0};  // Not strictly needed to zero the storage.
+        if (payload_size > 0U)                 // The check is needed to avoid calling memcpy() with a NULL pointer.
+        {
+            // Clang-Tidy raises an error recommending the use of memcpy_s() instead.
+            // We ignore it because the safe functions are poorly supported; reliance on them may limit the portability.
+            (void) memcpy(&scratch_data[0], payload, payload_size);  // NOLINT
+        }
+
+        mb->TDTR = payload_size;  // DLC equals data length except in CAN FD
+
+        mb->TDHR = (((uint32_t) scratch_data[7]) << 24U) | (((uint32_t) scratch_data[6]) << 16U) |
+                   (((uint32_t) scratch_data[5]) << 8U) | (((uint32_t) scratch_data[4]) << 0U);
+        mb->TDLR = (((uint32_t) scratch_data[3]) << 24U) | (((uint32_t) scratch_data[2]) << 16U) |
+                   (((uint32_t) scratch_data[1]) << 8U) | (((uint32_t) scratch_data[0]) << 0U);
```

Comment:
Can you please add `//` after `|` to make it nicely formatted?

Reply:
Done.